### PR TITLE
convert java to kotlin (10 last files within 'AndroidSDK..test')

### PR DESCRIPTION
**Pull request by abebeos**
Android Studio conversion functionality.

Manual changes explained with "// J2K:".

(re #1922)

### CODE REVIEW

**Comment by abebeos** [comment]
I just realized that the github UI does not help with the review of such conversions. Anyone has a hint on how to handle this, thus review is easy-as-usual?

**Comment by abebeos** [comment]
"convert kotlin to java" -> great, I reversed it in the commit message. So, new PR anyways.

**Comment by abebeos** [comment]
(@louwers , please use `rebase` if you must(!) update my work branch from the github UI. Those merge commits on my ongoing-work-branches confuse me terribly. I've update the branch via local cleanup&force-push + githut-UI-rebase)

**Comment by abebeos** [comment]
(last commit is meant "squash: j2k", will be squashed before merge)

Could you switch the CI to auto, thus I can see that all is good before going on with the next step?

**Comment by louwers** [comment]
@abebeos CI needs manual approval for first-time contributors, so for your next PR this will be the case.

Please use a 'Draft' PR while you are still working on it.

**Comment by abebeos** [comment]
> CI needs manual approval for first-time contributors, so for your next PR this will be the case.

@louwers , please go on with review/merge, I'll use a new PR for the next chunks.

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/LocationEngineRequestTest.kt
Comment (louwers) [inline platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/LocationEngineRequestTest.kt:13]: Can't you import `assertThat` like in Java? This is very verbose.

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt
Comment (louwers) [inline platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt:125]: Using a non-null assertion operator on `null` will throw a null pointer exception, but it won't even reach into the method, so you need a different way to pass `null` here.

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt
Comment (louwers) [inline platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt:73]: Here and in other places: is the cast needed?

If not, can you remove the explicit type annotation?

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt
Comment (louwers) [inline platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt:78]: This gets quite long and unreadable. I recommend to keep the imports of the static classes like in Java.

**Review by louwers — CHANGES_REQUESTED** [review]
Thanks for this PR!

Please see my comments. Other notes:

- The J2K comments are not needed, because once this is merged they are not relevant anymore.
- See if it is possible to get rid of some of the not-null assertions.
- Consider introducing Mockito Kotlin: https://github.com/mockito/mockito-kotlin

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/LocationEngineRequestTest.kt
Comment (abebeos) [reply platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/LocationEngineRequestTest.kt:13]: see https://github.com/maplibre/maplibre-native/pull/1926#issuecomment-1847334143

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt
Comment (abebeos) [reply platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt:125]: I spend the most time here to lookup for a solution, did not find something specific.

**Comment by louwers** [comment]
I don't see the benefit of doing multiple passes. Of course we should avoid making too big of a refactors in one go, but at least the static imports should remain as they exist in Java to avoid too much verbosity making the tests hard to read.

**Comment by abebeos** [comment]
(as a sidenote: some of the java static imports feel "too much" - but this is of course a matter of preference)

> I don't see the benefit of doing multiple passes. 

one is:

> avoid making too big of a refactors in one go

1st step is to take the machine conversion as-is, and only change things minimal to pass the tests. Beautifying is for 2nd step.

Adding the changes you mention is trivial, the thing is that you intervene in my process (and I have possibly a hundred of files more to process without 'loosing it'). My stomach revolts when I try to break standard-procedure (e.g. the "don't touch the machine conversion").

The other thing is that the existent J2K conversions seem to be "machine conversion 1st pass", too, see e.g.:

https://github.com/maplibre/maplibre-native/blob/main/platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/CompassEngineTest.kt

So I would prefer to go on like this for now (1st pass = machine conversion + minimum manual fixes to pass tests).

**Comment by abebeos** [comment]
@louwers , i understand your problems with the static imports, it really looks strange.

And Android Studio should(!) keep the static imports. It seems there is no option to set this.

But then, my insistence on strictness (machine conversion only) is valid , too.

I tried "IntellJ IDEA Community Edition", and it looks like the machine conversion keeps(!) the import statements, resulting in more compact code.

So, for now, it looks like we can add a simple requirement to the task "Java to Kotlin Conversion":

=> use a machine conversion that keeps the imports (like `import org.mockito.Mockito.doAnswer`), thus the resulting code remains compact. Example: IntelliJ IDEA Community Edition.

**Comment by abebeos** [comment]
Forget my last comment, intelliJ IDEA behaves the same, `import static` is not converted.

I see to it to find a way to do this with low effort, the resulting code starts to annoy me too (clearly a flaw in IntelliJ/AStudio J2K conversions)

**Comment by abebeos** [comment]
see related issue:

- #1938

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/LocationEngineRequestTest.kt
Comment (abebeos) [reply platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/LocationEngineRequestTest.kt]: imported

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt
Comment (abebeos) [reply platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt]: a code-err was flagged here by the IDE, followed the suggestions, the cast finally worked.

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt
Comment (abebeos) [reply platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt]: readded the (by machine-conversion dropped) imports

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt
Comment (louwers) [reply platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt:125]: OK I suggest we remove this test then, because the NullPointerException does not come from requestLocationUpdates right now.

**Comment by louwers** [comment]
Let me know if this is ready for review.

Note that the Kotlin style guide forbids wildcard imports:

https://developer.android.com/kotlin/style-guide

File: platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt
Comment (abebeos) [reply platform/android/MapboxGLAndroidSDK/src/test/java/org/maplibre/android/location/engine/MapLibreFusedLocationEngineImplTest.kt:127]: disabled them and added a TODO + Issue so this gets not forgotten. I have the feeling that if a library is used from java, then the null-ptr tests need to be in Java - but I'm not sure about it. 
98e4632

**Comment by abebeos** [comment]
> Let me know if this is ready for review again.

@louwers, should be ok now.

> Note that the Kotlin style guide forbids wildcard imports:

Google "forbids", but does not enforce the rules by its own tools (I added the dropped imports via the Android Studio IDE suggestions).

If this rule is important for the maplibre project, it could be possibly enforced by an Android Studio configuration.

**Comment by louwers** [comment]
You are right, this should be enforced on CI.

**Review by louwers — APPROVED** [review]
LGTM, maybe @artakka is able to have a look over it as well.

If not I will merge this tomorrow. Thanks for your contribution!

**Comment by abebeos** [comment]
> If you run `./gradlew ktlint` a lot of files get changed. I'll have a look at that.

The Android Studio has many warnings for the kotlin files (existent ones, prior to my pr), but it looks like that those should be process in one pass after the conversion.


As for Google / import .* :

* https://issuetracker.google.com/issues/205057065
* https://doc-26a0-27jg-issuetracker.googleusercontent.com/attachments/kc7eu567pksi0u9a1bnf6bomkfbp1d7c/bv7hgvia7kp5ltr7loa7v6tsv2dqnts4/1702407600000/192750/*/205057065::35378696::192750?q=AN9yeQLCLxWYx6mBKYIi5KRY3rV_ljbWvJfp4bFBfyedp6R7ThSBJAkjJLo8E_R6dzDYmK2bPcSqdIpuBOcNt6-xiQ&download=false

So, changing the configuration to "when at least 500" names used should stop the `.*` imports during conversion.

**Pull request merged**
